pkg/db/queries.sql

```diff
@@ -13,23 +13,36 @@ WHERE
 	singleton_id = 1;
 
 -- name: InsertGatewayEnvelope :execrows
-SELECT
-	insert_gateway_envelope(@originator_node_id, @originator_sequence_id, @topic, @originator_envelope);
+INSERT INTO gateway_envelopes(originator_node_id, originator_sequence_id, topic, originator_envelope)
+	VALUES (@originator_node_id, @originator_sequence_id, @topic, @originator_envelope)
+ON CONFLICT
+	DO NOTHING;
 
 -- name: SelectGatewayEnvelopes :many
+WITH cursors AS (
+	SELECT
+		UNNEST(@cursor_node_ids::INT[]) AS cursor_node_id,
+		UNNEST(@cursor_sequence_ids::BIGINT[]) AS cursor_sequence_id
+)
 SELECT
-	*
+	gateway_envelopes.*
 FROM
 	gateway_envelopes
+	-- Assumption: There is only one cursor per node ID. Caller must verify this
+	LEFT JOIN cursors ON gateway_envelopes.originator_node_id = cursors.cursor_node_id
 WHERE (sqlc.narg('topic')::BYTEA IS NULL
 	OR length(@topic) = 0
 	OR topic = @topic)
 AND (sqlc.narg('originator_node_id')::INT IS NULL
 	OR originator_node_id = @originator_node_id)
-AND (sqlc.narg('originator_sequence_id')::BIGINT IS NULL
-	OR originator_sequence_id > @originator_sequence_id)
-AND (sqlc.narg('gateway_sequence_id')::BIGINT IS NULL
-	OR id > @gateway_sequence_id)
+AND (cursor_sequence_id IS NULL
+	OR originator_sequence_id > cursor_sequence_id)
+ORDER BY
+	-- Assumption: envelopes are inserted in sequence_id order per originator, therefore
+	-- gateway_time preserves sequence_id order
+	gateway_time,
+	originator_node_id,
+	originator_sequence_id ASC
```

Comment:
Do we actually need these secondary sorts. Cheaper to index if it's just one field, and it doesn't seem to matter given that ties will be rare.

Reply:
I worry that consecutive payloads from the same originator could have the same timestamp. Maybe when we analyze the query performance later we can remove it if it's costly? Should be an easy change

Reply:
Is there an advantage to doing it this way vs using the table's `tid` as a tiebreaker (the Postgres default in ties)?

https://www.postgresql.org/docs/current/indexes-ordering.html

Reply:
I'm not sure to be honest. This is the definition of `tid` I found:

```
A final identifier type used by the system is tid, or tuple identifier (row identifier). This is the data type of the system column ctid. A tuple ID is a pair (block number, tuple index within block) that identifies the physical location of the row within its table.
```

It seems somewhat tied to the implementation. I'm inclined to write the query first in the naive/intuitive way, and determine later on whether or not there is a performance impact - I'm somewhat hopeful that the query planner will optimize it seeing as we won't have ties most of the time.

Reply:
Open to feedback though! Definitely not my expertise.

Reply:
Query planner is pretty smart. You might be right it'll just use the index on the first column in the ORDER BY and reshuffle in memory for ties.

We can find out when we get to optimizations and profiling